core/src/Kokkos_View.hpp

```diff
@@ -1811,13 +1812,20 @@ KOKKOS_INLINE_FUNCTION bool operator!=(const View<LT, LP...>& lhs,
 
 namespace Kokkos {
 namespace Impl {
+struct SharedAllocationDisableTrackingGuard {
```

Comment:
Guard classes are normally neither moveable nor copyable.
```suggestion
struct SharedAllocationDisableTrackingGuard {
  SharedAllocationDisableTrackingGuard(SharedAllocationDisableTrackingGuard const&) = delete;     
  SharedAllocationDisableTrackingGuard& operator=(SharedAllocationDisableTrackingGuard const&) = delete;
  SharedAllocationDisableTrackingGuard(SharedAllocationDisableTrackingGuard&&) = delete;     
  SharedAllocationDisableTrackingGuard& operator=(SharedAllocationDisableTrackingGuard&&) = delete;
  ~SharedAllocationDisableTrackingGuard() = default;
  // ...
```

I'm not thrilled with this class, because the destructor always enables tracking, even if it was disabled when it was constructed.  That means we have to be careful when using it, such as making sure it is never used in a nested way.  Maybe it should be a precondition (enforced with an assert) that tracking is enabled before construction, and a precondition that tracking is disabled before destruction?  FYI: I'd be fine with only an assert on construction.

You could do something more complicated such as keeping track of the initial enabled state and doing nothing if it wasn't enabled in the first place, but given this is internal Kokkos use only, that seems like overkill.

Reply:
I deleted the move/copy constructors/assignment operators and asserted in the constructor